src/coreclr/vm/pendingload.h

```diff
@@ -17,248 +17,404 @@
+        {
+            CONTRACTL
+            {
+                THROWS;
+                GC_TRIGGERS;
+                INJECT_FAULT(COMPlusThrowOM(););
+            }
+            CONTRACTL_END;
+
+            if (m_pException)
+                PAL_CPP_THROW(Exception *, m_pException->Clone());
+
+            _ASSERTE(FAILED(m_hrResult));
+
+            if (m_hrResult == COR_E_TYPELOAD)
+            {
+                TypeKey typeKey = GetTypeKey();
+                ClassLoader::ThrowTypeLoadException(&typeKey,
+                                                    IDS_CLASSLOAD_GENERAL);
+
+            }
+            else
+                EX_THROW(EEMessageException, (m_hrResult));
         }
-        else
-            EX_THROW(EEMessageException, (m_hrResult));
-    }
 
-    void SetException(Exception *pException)
-    {
-        CONTRACTL
+        void SetException(Exception *pException)
         {
-              NOTHROW;
-              PRECONDITION(HasLock());
-              PRECONDITION(m_pException == NULL);
-              PRECONDITION(m_dwWaitCount > 0);
+            CONTRACTL
+            {
+                NOTHROW;
+                PRECONDITION(HasLock());
+                PRECONDITION(m_pException == NULL);
+                PRECONDITION(m_dwWaitCount > 0);
+            }
+            CONTRACTL_END;
+
+            m_typeHandle = TypeHandle();
+            m_hrResult = COR_E_TYPELOAD;
+
+            // we don't care if this fails
+            // we already know the HRESULT so if we can't store
+            // the details - so be it
+            EX_TRY
+            {
+                FAULT_NOT_FATAL();
+                m_pException = pException->Clone();
+            }
+            EX_CATCH
+            {
+                m_pException=NULL;
+            }
+            EX_END_CATCH(SwallowAllExceptions);
         }
-        CONTRACTL_END;
 
-        m_typeHandle = TypeHandle();
-        m_hrResult = COR_E_TYPELOAD;
+        void SetResult(TypeHandle typeHnd)
+        {
+            CONTRACTL
+            {
+                NOTHROW;
+                PRECONDITION(HasLock());
+                PRECONDITION(m_pException == NULL);
+                PRECONDITION(m_dwWaitCount > 0);
+            }
+            CONTRACTL_END;
+
+            m_typeHandle = typeHnd;
+        }
 
-        // we don't care if this fails
-        // we already know the HRESULT so if we can't store
-        // the details - so be it
-        EX_TRY
+        void UnblockWaiters()
         {
-            FAULT_NOT_FATAL();
-            m_pException = pException->Clone();
+            CONTRACTL
+            {
+                NOTHROW;
+                PRECONDITION(HasLock());
+                PRECONDITION(m_dwWaitCount > 0);
+            }
+            CONTRACTL_END;
+
+            _ASSERTE(m_fLockAcquired);
+            m_Crst.Leave();
+            m_fLockAcquired = FALSE;
         }
-        EX_CATCH
+    #endif //DACCESS_COMPILE
+
+        TypeKey& GetTypeKey()
```

Comment:
Is mutability desired here?

Reply:
No, we just want the type key. I've fixed this by making it a const pointer, and then updating the various parts of our codebase to allow a const TypeKey pointer to be used. Looks a little better to me.